flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java

```diff
@@ -181,6 +181,14 @@ void retriggerSubpartitionRequest(int subpartitionIndex) throws IOException {
 			moreAvailable = !receivedBuffers.isEmpty();
 		}
 
+		if (next == null) {
```

Comment:
I guess it's theoretically impossible that we get a null buffer here with your changes in `releaseAllResources`, which seems to solve two cases you mentioned in description. So.. this check is just for other unknown bad cases?

Reply:
I guess it can probably happen in practice. When the canceler thread already released the respective input channel, but the task thread might still call `getNextBuffer` in the case of released `receivedBuffers`, then it can get the `null` buffer.

Reply:
What do you mean @zhijiangW ? At first glance I would agree with @Jiayi-Liao, that it shouldn't happen after your fix in this commit in this class below.

Reply:
Let me further explain it. 

The canceler thread will close the `InputGate` in advance so the task thread might be aware of the released state to exit early. So when the canceler thread called `RemoteInputChannel#releaseAllResources` before, then all the buffers in `receivedBuffers` were already drained and recycled. 

But the task thread was not aware of this then, and it would probably call `getNextBuffer` to get a `null` buffer here. We only expect the `null` buffer in the case of `released` channel, so we throw the expected `CancelTaskException` to make task thread exit. If not released case, there must be some logic bugs. E.g this channel notifies gate of available data by fault.  So we throw `IllegalStateException` for such case, to avoid the misleading `NullPointerException` when reference with the buffer below.

My fix in `#releaseAllResources` is only for avoiding concurrent pulling `receivedBuffers` by both task thread and canceler thread, which might cause recycle the same buffer twice and misleading exception thrown by netty stack.  

Another option for modifying the logic in `#getNextBuffer` like below: 

```
synchronized (receivedBuffers) {
			if (isReleased.get()) {
				throw new CancelTaskException("Queried for a buffer after channel has been released.");
			}
			next = receivedBuffers.poll();
			moreAvailable = !receivedBuffers.isEmpty();
		}
```

But it might still make sense to also judge whether the buffer is `null` out of the `synchronized`, which is not the race condition case, and only for avoiding potential logic bugs in data notification logic to cause misleading NPE.

Reply:
FYI: before my changes, we can reproduce these two potential issues in my below introduced unit tests stable. One case is throwing NPE exception sometimes, and another case is throwing netty stack exception when same buffer recycled twice.

Reply:
Get it! Thanks for the explaination!

Reply:
Ops, sorry, yes it's obvious, I don't know why I've missed it 😳